Docs for Sourcegraph Architecture (#1067)
sourcegraph/docs: docs/admin/architecture.mdx

## Diagram

![sourcegraph-architecture](https://storage.googleapis.com/sourcegraph-assets/Docs/sg-architecture.svg)

## Repository syncing

At its core, Sourcegraph maintains a persistent cache of all repositories that are connected to it. It is persistent because this data is critical for Sourcegraph to function. Still, it is ultimately a cache because the code host is the source of truth, and our cache is eventually consistent.

- `gitserver` is the sharded service that stores repositories and makes them accessible to other Sourcegraph services
- `repo-updater` is the singleton service responsible for ensuring all repositories in gitserver are as up-to-date as possible while respecting code host rate limits. It is also responsible for syncing repository metadata from the code host that is stored in the repo table of our Postgres database

## Permission syncing

Repository permissions are mirrored from code hosts to Sourcegraph by default. This builds the foundation of Sourcegraph authorization for repositories to ensure users see consistent content on code hosts. Currently, the background permissions syncer resides in the repo-updater.

> To learn more about how repository permissions are synchronized in the background, read about [permission syncing](/admin/permissions/syncing).

Reviewer: ```suggestion
<Callout type="note">Learn more in the [Permission Syncing docs](/admin/permissions/syncing)</Callout>
```

## Search

Developers can search for the entire codebase that is connected to their Sourcegraph instance.

By default, Sourcegraph uses `zoekt` to create a trigram index of the default branch of every repository, which makes searches fast. This trigram index is why Sourcegraph search is more powerful and faster than what is usually provided by code hosts.

- [zoekt-indexserver](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/tree/cmd/zoekt-sourcegraph-indexserver)
- [zoekt-webserver](https://sourcegraph.com/github.com/sourcegraph/zoekt/-/tree/cmd/zoekt-webserver)

Sourcegraph also has a fast search path for code that isn't indexed yet or will never be indexed (for example, code that is not on a default branch). Indexing every branch of every repository isn't a pragmatic use of resources for most customers, so this decision balances optimizing the common case (searching all default branches) with space savings (not indexing everything).

- `searcher` implements the non-indexed search
- Syntax highlighting for any code view, including search results, is provided by `Syntect` server

> You can learn more about search in the [Code Search docs](/code-search).

Reviewer: ```suggestion
<Callout type="note">Learn more in the [Code Search docs](/code-search)</Callout>
```

## Code Navigation

> Code Navigation surfaces data (for example, doc comments for a symbol) and actions (for example, go to definition, find references) based on our semantic understanding of code (unlike search, which is completely text-based).

Reviewer: suggestion to avoid parentheses overload:

```suggestion
Unlike Search (which is completely text-based), Code Navigation surfaces data such as doc comments for a symbol and actions such as the "go to definition" or "find references" features based on our semantic understanding of code.
```

By default, Sourcegraph provides [search-based code navigation](/code-search/code-navigation/search_based_code_navigation). This reuses all the architecture that makes search fast, but it can result in false positives (for example, finding two definitions for a symbol or references that aren't actually references) or false negatives (for example, not being able to find the definition or all references).

> 

Reviewer: No reference to our syntactic-indexing solution here. I don't know how explicitly we want to document this new (beta?) feature but it should probably be mentioned for completeness.

Author: @loujar, just for confirmation, are you referencing the auto-indexing feature?

This is the default because it works with no extra configuration and is good for many use cases and languages. We support many languages this way because it only requires writing a few regular expressions.

With some setup, customers can enable [precise code navigation](/code-search/code-navigation/precise_code_navigation). Repositories add a step to their build pipeline that computes the index for that code revision and uploads it to Sourcegraph. We must write language-specific indexers, so adding precise code navigation support for new languages is a non-trivial task.

> To learn more about [Code Navigation, read the docs here](/code-search/code-navigation).

Reviewer: ```suggestion
<Callout type="note">Learn more in the [Code Navigation docs](/code-search/code-navigation)</Callout>
```

Once created, Sourcegraph monitors changesets, and their current review and CI status can be viewed on the batch change page. This provides a single pane of glass view of all the changesets created as part of the batch change. The batch change can be updated at any time by re-applying the original batch spec: this will transparently add or remove changesets in repositories that now match or don't match the original search as needed.

Read the [Batch Changes](/batch-changes) docs to learn more.

### Dependencies

- src-cli: Batch changes are currently executed client-side through the `src` CLI
- Search: Repositories in which batch specs need to be executed are resolved through the search API

> ## Code insights

Reviewer: noting inconsistent capitalization patterns across our header section text, but not sure what the correct style is

> Code insights surface higher-level, aggregated information to leaders in engineering organizations in dashboards. For example, code insights can track the number of matches of a search query over time, the number of code navigation diagnostic warnings in a codebase, or the usage of different programming languages. Sample use cases for this are tracking migrations, the usage of libraries across an organization, tech debt, code base health, and much more.

Reviewer: greater question regarding the content of this doc; should we maintain detailed feature descriptions throughout this page? I personally feel that we should link to the respective feature pages for these topics, and have the content focus strictly on the architectural details of the feature.

Reviewer 2: I have also been wondering this. Do we think a detailed description of the individual services would be more appropriate, and can reference the impact that has on specific features if applicable?

  - Insights contributed by extensions are configured through settings (this will stay the same)

## Code Monitoring

Code Monitoring allows users to get notified of changes to their codebase.

Users can view, edit, and create code monitors through the code monitoring UI. A code monitor comprises a trigger and one or more actions.

The **trigger** watches for new data; if there is new data, we call this an event. The only supported trigger is a search query of `type:diff` or `type:commit`, run every five minutes by the Go backend with an automatically added `after:` parameter narrowing down the diffs/commits that should be searched. The monitor's configured actions are run when this query returns a non-zero number of results.

The actions are run in response to a trigger event. For now, the only supported action is an email notification to the primary email address of the code monitor's owner. To work, `email.address` and `email.smtp` must be configured in the site configuration. Code monitoring actions will be extended in the future to support webhooks.

> Learn more in the [Code Monitoring](/code_monitoring) docs.

Reviewer: ```suggestion
<Callout type="note">Learn more in the [Code Monitoring docs](/code_monitoring)</Callout>
```

### Dependencies

- Search: Diff and commit search triggers

## Browser extensions

The [Sourcegraph browser extensions](/integration/browser_extension) bring the features of Sourcegraph directly into the UI of code hosts such as GitHub, GitLab, and Bitbucket.

> With the Sourcegraph browser extension installed, users get Sourcegraph features (including code navigation) on their code host while browsing code, viewing diffs, or reviewing pull requests.

Reviewer: ```suggestion
With the Sourcegraph browser extension installed, users get Sourcegraph features (including Code Navigation) on their code host while browsing code, viewing diffs, or reviewing pull requests.
```

This lets users get value from Sourcegraph without leaving existing workflows on their code host. It also gives them a convenient way to jump into Sourcegraph anytime (by using the **Open in Sourcegraph** button on any repository or file). The browser extension also adds an address bar search shortcut, allowing you to search on Sourcegraph directly from the browser address bar.

## Native integrations (for code hosts)

Native integrations bring Sourcegraph features directly into the UI of code hosts, similar to browser extensions.

Instead of requiring a browser extension, native integrations inject a script by extending the code host directly (for example, using the code host's plugin architecture). The advantage is that Sourcegraph can be enabled for all users of a code host instance without any action required from each user.

> Learn more about our [Code host integrations in the docs](/admin/code_hosts).

Reviewer: ```suggestion
<Callout type="note">Learn more in the [Code host integrations docs](/admin/code_hosts)</Callout>
```

#### Dependencies

- Repository Syncing: Uses the GraphQL API to resolve repositories and revisions on code hosts
- Search: Query transformer API hooks into search in the web app
- Settings cascade: Which extensions are enabled, and which configurations for extensions are stored in the settings. Extensions may also change settings

## src-cli

`src-cli`, or `src`, is a command line tool that users can run locally to interact with Sourcegraph.

`src-cli` is written in Go and distributed as a standalone binary for Windows, macOS, and Linux. Its features include running searches, managing Sourcegraph, and executing batch changes. `src-cli` is integral to the batch changes product.

> Note that `src-cli` is not contained within the Sourcegraph monorepo and has its release cadence.

Reviewer: ```suggestion
<Callout type="note">`src-cli` is a standalone client, maintained and released separately from Sourcegraph</Callout>
```

> Learn more about `src-cli` in the [docs](/cli) and [GitHub repo](https://github.com/sourcegraph/src-cli).

Reviewer: ```suggestion
<Callout type="note">Learn more in the [src-cli docs](/admin/code_hosts) and [GitHub repo](https://github.com/sourcegraph/src-cli)</Callout>
```

### Dependencies

- Search: GraphQL API
- Batch Changes: GraphQL API

> 

Reviewer: ```suggestion
- Code Intelligence: GraphQL API

```

## Deployment

> Sourcegraph is deployable via three supported methods:

Reviewer: This list is not up to date. @cloudyadi could you revise this section?

- **[Kubernetes](/admin/deploy/kubernetes)** is intended for all medium to large-scale production deployments that require fault tolerance and high availability. For advanced users only with significant kubernetes experience required. This deployment method is developed in deploy-sourcegraph
- **[Docker Compose](/admin/deploy/docker-compose)** is intended to be used for small to medium production deployments, with some customization available. Easy to set up with basic infrastructure and docker knowledge required. A variation on this is the pure-Docker option. Both of these deployment methods are developed in deploy-sourcegraph-docker

> - **[Docker Single Container](/admin/deploy/docker-single-container)** for small environments on a single server. It is the easiest and quickest to set up with a single command. Little infrastructure knowledge is required. This deployment method is developed in cmd/server

Reviewer: ```suggestion
```

The [resource estimator](/admin/deploy/resource_estimator#sourcegraph-resource-estimator) can guide you on the requirements for each deployment type.

## Observability

Observability encapsulates the monitoring and debugging of Sourcegraph deployments. Sourcegraph is designed and ships several observability tools and out-of-the-box capabilities to enable visibility into the health and state of a Sourcegraph deployment.

Monitoring includes [metrics and dashboards](/admin/observability/metrics), [alerting](/admin/observability/alerting), and [health checking](/admin/observability/health_checks) capabilities.

> - grafana is the frontend for service metrics and ships with customized dashboards for Sourcegraph services

Reviewer: another general style note, we inconsistently use `code-formatting` for container service names through this doc

- prometheus handles the scraping of service metrics and ships with recording rules, alert rules, and alerting capabilities
- cadvisor provides per-container performance metrics (scraped by Prometheus) in most Sourcegraph environments
- Each Sourcegraph service provides health checks

Debugging includes [tracing](/admin/observability/tracing) and [logging](/admin/observability/logs).

- jaeger is the distributed tracing service used by Sourcegraph

> - jaeger-agent is a sidecar used in Kubernetes deployments for collecting traces from services

Reviewer: My understanding is we have fully removed the `jaeger-agent` service
https://sourcegraph.slack.com/archives/C02E4HE42BX/p1743419855806249

Author: Thanks for the catch. I'll remove it's mention.

- Each Sourcegraph service provides logs

> Learn more in our [Observability docs](/admin/observability).

Reviewer: <Callout type="note">Learn more in the [Observability docs](/admin/observability)</Callout>
